### PR TITLE
set cilium agent only run on linux nodes

### LINKED ISSUE #14410: cilium-agent gets scheduled to nodes running OSes other than Linux
## Bug report

[`cilium-agent` is deployed through a DaemonSet](https://github.com/cilium/cilium/blob/master/install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml). Currently, there doesn't seem to be a way for a cluster admin to ensure it only gets scheduled onto Linux nodes. Why?
- DaemonSet doesn't set a `nodeSelector` to avoid OSes other than Linux, ie
  ```yaml
  nodeSelector:
      kubernetes.io/os: linux
  ```
  FYI the kubelet sets the [`kubernetes.io/os`](https://kubernetes.io/docs/reference/kubernetes-api/labels-annotations-taints/#kubernetes-io-os) node label, based on `GOOS`.
- DaemonSet [specifies a catch-all toleration that invalidates the usage of taints](https://github.com/cilium/cilium/blob/5b911b80127d7d79120252334b92f7fd6495831c/install/kubernetes/cilium/values.yaml#L123-L124).

**General Information**

- Cilium version (run `cilium version`): 1.8.4
- Kernel version (run `uname -a`): `uname -a` doesn't work on Windows
- Orchestration system version in use (e.g. `kubectl version`, Mesos, ...): Kubernetes 1.19.x
- Link to relevant artifacts (policies, deployments scripts, ...): all linked above.
- Generate and upload a system zip: doesn't seem relevant and actually I can't do it.
- Relevant slack thread: https://cilium.slack.com/archives/C53TG4J4R/p1607980778460600

**How to reproduce the issue**

1. Add a Windows node to a Kubernetes cluster and immediately cilium-agent is deployed there, obviously failing to run.
2. Alternatively, deploy a virtual-kubelet provider (eg [AWS Fargate](https://github.com/virtual-kubelet/aws-fargate)) and immediately cilium-agent is deployed there, when it shouldn't.))

### PR DESCRIPTION
**Pull request by answer1991**
Fixes: #14410

```release-note
set cilium agent only run on linux nodes
```

### CODE REVIEW

File: install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml [hunk 1]
```diff
@@ -58,6 +58,8 @@ spec:
 {{ toYaml .Values.affinity | indent 8 }}
 {{- end }}
 {{- if .Values.imagePullSecrets }}
+      nodeSelector:
```
Comment (sayboras) [inline install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml]: should this `nodeSelector` be independent with `imagePullSecrets` ?

**Review by sayboras — COMMENTED** [review]
Thanks for your PR :100:, I have one comment as per below.

FYI. `make -C install/kubernetes all` can be used to lint and generate related files.

File: install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml (hunk 1 above)
Comment (pires) [reply install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml]: I agree it should.

Comment (answer1991) [reply install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml]: yes, I made a mistake and I have already fixed it, thanks!

**Comment by ungureanuvladvictor** [comment]
test-me-please

**Comment by ungureanuvladvictor** [comment]
@answer1991 in k8s 1.14.x the labels OS + arch labels were promoted to GA (https://github.com/kubernetes/kubernetes/pull/73048). We run our e2e tests on k8s 1.13.x as well where they were still marked with `beta.` this is why `K8s-1.13-Kernel-netnext` failed.

Per https://docs.cilium.io/en/v1.9/concepts/kubernetes/compatibility/#kubernetes-compatibility we support k8s 1.12.x -> 1.19.x so maybe add the other label deprecated label selector and we remove it when we drop support for 1.12.x and 1.13.x ?

**Comment by answer1991** [comment]
> @answer1991 in k8s 1.14.x the labels OS + arch labels were promoted to GA ([kubernetes/kubernetes#73048](https://github.com/kubernetes/kubernetes/pull/73048)). We run our e2e tests on k8s 1.13.x as well where they were still marked with `beta.` this is why `K8s-1.13-Kernel-netnext` failed.
> 
> Per https://docs.cilium.io/en/v1.9/concepts/kubernetes/compatibility/#kubernetes-compatibility we support k8s 1.12.x -> 1.19.x so maybe add the other label deprecated label selector and we remove it when we drop support for 1.12.x and 1.13.x ?

Let me see if we can use `affinity` to solve it.

**Comment by answer1991** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Review by aanm — APPROVED** [review]

**Review by sayboras — APPROVED** [review]
LGTM :100:

**Comment by pires** [comment]
My understanding of node affinity says this is incorrect as both expressions must match. If I'm correct, this will work for 1.19, as both beta and stable labels are set but not for 1.13 where I think only the beta labels are set.

**Comment by pires** [comment]
Actually the beta labels are not set on 1.19 and newer https://github.com/kubernetes/kubernetes/commit/54c0f8b677d0b82258f3b4df6d325cc3c0011661#diff-f39bb531745e5658a65143c9054c3fa92aa3ac9f41e9e627ac61b7ff08808153. Maybe tooling like kubeadm is setting it though because I'm seeing it in a test 1.19.x cluster and these tests passed for 1.19.

**Comment by answer1991** [comment]
> My understanding of node affinity says this is incorrect as both expressions must match

No, the NodeSelectorTerms are ORed.

**Comment by answer1991** [comment]
I have no idea about why the test failed. Is the test flaky?

**Comment by pires** [comment]
You're right, thanks. Relevant documentation https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#nodeselector-v1-core.

**Comment by aanm** [comment]
retest-net-next

**Comment by aanm** [comment]
retest-4.9

**Comment by aanm** [comment]
previous failures:
https://jenkins.cilium.io/job/Cilium-PR-K8s-1.13-net-next/297/testReport/Suite-k8s-1/13/K8sServicesTest_Checks_service_across_nodes_Tests_NodePort_BPF_Tests_with_direct_routing_Tests_LoadBalancer_Connectivity_to_endpoint_via_LB/
https://jenkins.cilium.io/job/Cilium-PR-K8s-1.20-kernel-4.9/271/testReport/junit/Suite-k8s-1/20/K8sServicesTest_Checks_ClusterIP_Connectivity_Checks_service_kubernetes_io_service_proxy_name_label_implementation/

**Pull request merged**